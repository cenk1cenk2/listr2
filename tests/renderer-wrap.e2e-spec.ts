/* eslint-disable @typescript-eslint/no-empty-function */
import delay from 'delay'
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root/index'

describe('fallback renderer', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  const message =
    // eslint-disable-next-line max-len
    'THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE'

  const messageWithNewLines =
    // eslint-disable-next-line max-len
    'THIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE'

  const messageWithEmptyLines =
    // eslint-disable-next-line max-len
    'THIS IS A LONG LONG MESSAGE\n\n\nAGE\nTHIS ISESSAGE\nTHIS IG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE'

  process.stdout.isTTY = true
  process.stdout.columns = 80

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  it.each([
    [ 'truncate', message ],
    [ 'wrap', message ],
    [ 'truncate', messageWithNewLines ],
    [ 'wrap', messageWithNewLines ]
  ])('should %s long strings', async (format, m) => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: `This task will with formating the output with ${format} output.`,
            task: async (_, task): Promise<void> => {
              task.output = m
              await delay(20)
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, formatOutput: format as any }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeFalsy()
    expect(mockStdout.mock.calls).toMatchSnapshot()
    expect(mockStderr.mock.calls).toMatchSnapshot()
    expect(mockExit.mock.calls).toMatchSnapshot()
  })

  it.each([
    [ 'truncate', message ],
    [ 'wrap', message ],
    [ 'truncate', messageWithNewLines ],
    [ 'wrap', messageWithNewLines ]
  ])('should %s long titles', async (format, m) => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: m,
            task: async (_, task): Promise<void> => {
              task.output = `This task will with formating the title with ${format} output.`
              await delay(20)
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, formatOutput: format as any }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeFalsy()
    expect(mockStdout.mock.calls).toMatchSnapshot()
    expect(mockStderr.mock.calls).toMatchSnapshot()
    expect(mockExit.mock.calls).toMatchSnapshot()
  })

  it.each([
    [ true, 'truncate', messageWithEmptyLines ],
    [ false, 'truncate', messageWithEmptyLines ]
  ])('should skip empty lines %s', async (rel, format, m) => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: `This will skip empty lines on output: ${rel}`,
            task: async (_, task): Promise<void> => {
              task.output = m
              await delay(20)
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: {
            lazy: true,
            formatOutput: format as any,
            removeEmptyLines: rel
          }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeFalsy()
    expect(mockStdout.mock.calls).toMatchSnapshot()
    expect(mockStderr.mock.calls).toMatchSnapshot()
    expect(mockExit.mock.calls).toMatchSnapshot()
  })
})
