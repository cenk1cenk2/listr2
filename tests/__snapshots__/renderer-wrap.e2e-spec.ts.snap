// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fallback renderer should skip empty lines false 1`] = `Array []`;

exports[`fallback renderer should skip empty lines false 2`] = `Array []`;

exports[`fallback renderer should skip empty lines false 3`] = `Array []`;

exports[`fallback renderer should skip empty lines true 1`] = `Array []`;

exports[`fallback renderer should skip empty lines true 2`] = `Array []`;

exports[`fallback renderer should skip empty lines true 3`] = `Array []`;

exports[`fallback renderer should truncate long strings 1`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "‚óº This task will with formating the output with truncate output.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will with formating the output with truncate output.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ This task will with formating the output with truncate output.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ This task will with formating the output with truncate output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LO‚Ä¶
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ This task will with formating the output with truncate output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LO‚Ä¶
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚úî This task will with formating the output with truncate output.
",
  ],
  Array [
    "[?25h",
  ],
]
`;

exports[`fallback renderer should truncate long strings 2`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "[?25h",
  ],
]
`;

exports[`fallback renderer should truncate long strings 3`] = `Array []`;

exports[`fallback renderer should truncate long strings 4`] = `Array []`;

exports[`fallback renderer should truncate long strings 5`] = `Array []`;

exports[`fallback renderer should truncate long strings 6`] = `Array []`;

exports[`fallback renderer should truncate long titles 1`] = `Array []`;

exports[`fallback renderer should truncate long titles 2`] = `Array []`;

exports[`fallback renderer should truncate long titles 3`] = `Array []`;

exports[`fallback renderer should truncate long titles 4`] = `Array []`;

exports[`fallback renderer should truncate long titles 5`] = `Array []`;

exports[`fallback renderer should truncate long titles 6`] = `Array []`;

exports[`fallback renderer should wrap long strings 1`] = `Array []`;

exports[`fallback renderer should wrap long strings 2`] = `Array []`;

exports[`fallback renderer should wrap long strings 3`] = `Array []`;

exports[`fallback renderer should wrap long strings 4`] = `Array []`;

exports[`fallback renderer should wrap long strings 5`] = `Array []`;

exports[`fallback renderer should wrap long strings 6`] = `Array []`;

exports[`fallback renderer should wrap long titles 1`] = `Array []`;

exports[`fallback renderer should wrap long titles 2`] = `Array []`;

exports[`fallback renderer should wrap long titles 3`] = `Array []`;

exports[`fallback renderer should wrap long titles 4`] = `Array []`;

exports[`fallback renderer should wrap long titles 5`] = `Array []`;

exports[`fallback renderer should wrap long titles 6`] = `Array []`;
